### PR TITLE
subsystem benchmarks: add cpu profiling

### PR DESCRIPTION
**Pull request by AndreiEres**
- Added optional CPU profiling
- Updated instructions how to set up Prometheus, Pyroscope and Graphana
- Added a flamegraph dashboard
image

### CODE REVIEW

File: polkadot/node/subsystem-bench/README.md [hunk 1]
```diff
@@ -26,17 +26,26 @@ The output binary will be placed in `target/testnet/subsystem-bench`.
...
 
-Subsystem, CPU usage and network  metrics are exposed via a prometheus endpoint during the test execution.
-A small subset of these collected metrics are displayed in the CLI, but for an in depth analysys of the test results, 
+Subsystem, CPU usage and network metrics are exposed via a prometheus endpoint during the test execution.
+A small subset of these collected metrics are displayed in the CLI, but for an in depth analysys of the test results,
 a local Grafana/Prometheus stack is needed.
 
+### Run Prometheus, Pyroscope and Graphana in Docker
+
+If you are using Docker, you can skip the next steps to install Prometheus, Pyroscope and Graphana using `docker-compose` file.
```
Comment (sandreim) [inline polkadot/node/subsystem-bench/README.md:35]:
```suggestion
If docker is not usable, then follow the next sections to manually install Prometheus, Pyroscope and Graphana on your machine.
```

File: polkadot/node/subsystem-bench/README.md [hunk 2]
```diff
@@ -26,17 +26,26 @@ The output binary will be placed in `target/testnet/subsystem-bench`.
...
+
+```bash
+cd polkadot/node/subsystem-bench/docker
+docker compose up
```
Comment (sandreim) [inline polkadot/node/subsystem-bench/README.md:39]: Nice!

File: polkadot/node/subsystem-bench/grafana/cpu-profiling.json [hunk 3]
```diff
@@ -0,0 +1,70 @@
+{
```
Comment (sandreim) [inline polkadot/node/subsystem-bench/grafana/cpu-profiling.json:1]: Is this a real time flamegraph or how does it work ? I would love to see a screenshot :)

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs [hunk 4]
```diff
@@ -76,12 +76,31 @@ struct BenchCli {
...
 
 impl BenchCli {
 	fn launch(self) -> eyre::Result<()> {
+		use pyroscope::PyroscopeAgent;
+		use pyroscope_pprofrs::{pprof_backend, PprofConfig};
+
+		// Pyroscope must be running on port 4040
+		// See https://grafana.com/docs/pyroscope/latest/get-started/#download-and-configure-pyroscope
+		let agent_running = if self.profile {
+			let agent = PyroscopeAgent::builder("http://localhost:4040", "subsystem-bench")
```
Comment (sandreim) [inline polkadot/node/subsystem-bench/src/subsystem-bench.rs:95]: To not couple the code to a specific deployment configuration this needs to be a configurable parameter of the CLI.

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs [hunk 5]
```diff
@@ -76,12 +76,31 @@ struct BenchCli {
...
+				.backend(pprof_backend(PprofConfig::new().sample_rate(100)))
```
Comment (sandreim) [inline polkadot/node/subsystem-bench/src/subsystem-bench.rs:96]: sample rate also should be exposed in CLI.

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs [hunk 6]
```diff
@@ -76,12 +76,31 @@ struct BenchCli {
 	/// Maximum remote peer latency in milliseconds [0-5000].
 	pub peer_max_latency: Option<u64>,
 
+	#[clap(long, default_value_t = false)]
+	/// Enable CPU Profiling
```
Comment (sandreim) [inline polkadot/node/subsystem-bench/src/subsystem-bench.rs:80]: This needs to be more explicit, we are using pyroscope.

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs [hunk 7]
```diff
@@ -76,12 +76,31 @@ struct BenchCli {
...
+	pub profile: bool,
+
 	#[command(subcommand)]
 	pub objective: cli::TestObjective,
 }
 
 impl BenchCli {
 	fn launch(self) -> eyre::Result<()> {
+		use pyroscope::PyroscopeAgent;
+		use pyroscope_pprofrs::{pprof_backend, PprofConfig};
```
Comment (sandreim) [inline polkadot/node/subsystem-bench/src/subsystem-bench.rs:90]: I would move these at top of file.

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs [hunk 8]
```diff
@@ -165,6 +184,11 @@ impl BenchCli {
 		env.runtime()
 			.block_on(availability::benchmark_availability_read(&mut env, state));
 
+		if let Some(agent_running) = agent_running {
```
Comment (sandreim) [inline polkadot/node/subsystem-bench/src/subsystem-bench.rs:187]: This code has changed, you should update the branch and make sure we still run it before exit.

File: polkadot/node/subsystem-bench/grafana/cpu-profiling.json (hunk 3 above)
Comment (AndreiEres) [reply polkadot/node/subsystem-bench/grafana/cpu-profiling.json:1]: Added to the description

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs (hunk 4 above)
Comment (AndreiEres) [reply polkadot/node/subsystem-bench/src/subsystem-bench.rs]: Ok, but with default to fast run it like we do with Prometheus

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs (hunk 6 above)
Comment (AndreiEres) [reply polkadot/node/subsystem-bench/src/subsystem-bench.rs]: Ok, but with default to fast run it like we do with Prometheus

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs (hunk 7 above)
Comment (AndreiEres) [reply polkadot/node/subsystem-bench/src/subsystem-bench.rs]: Ok

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs (hunk 8 above)
Comment (AndreiEres) [reply polkadot/node/subsystem-bench/src/subsystem-bench.rs:192]: Ok

**Comment by alindima** [comment]
I wonder, did you see any performance impact on the benchmark when running with pyroscope profiling enabled?

**Pull request closed without merging**